### ISSUE TITLE
Thread.Name should use SetThreadDescription

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/coreclr#12354]
@Alois-xx commented on [Sun Jun 18 2017](https://github.com/dotnet/corefx/issues/21190)

Windows 10 Creators Update delivers new APIs to set the thread name 

https://msdn.microsoft.com/en-us/library/windows/desktop/mt774976(v=vs.85).aspx

These should be used by the managed thread name set property as well so one can find the thread name in ETW traces on Windows as well. Currently there was no way to set the thread name for unmanaged threads with a decent API until Creators Update. 

---

@danmosemsft commented on [Sun Jun 18 2017](https://github.com/dotnet/corefx/issues/21190#issuecomment-309300246)

@mikem8361 @brianrob as I understand it, `Thread.Name` is just a contract between CLR and the debugger and the value is stored on Thread. The idea is it could also SetThreadDescription.

**Comment by danmosemsft** [comment]
@vancem can you comment on the profiler observation? is this something that's come up before? should Thread.Name also call SetThreadDescription()?

**Comment by vancem** [comment]
Obviously until this API exists, at the OS level of abstraction threads don't have names, and thus ETW profiling tools don't know about that name.   Using this API is one step in fixing that, but of course ETW tools would also have to be updated.   

I think it is certainly a reasonable thing to do (passing what information we have to the 'standard' the that OS defines.  It can only make things better (although it may take a while until the whole food chain takes advantage of it.   

We should also consider calling this name for worker threads that the runtime creates on its own (e.g. the debugger thread, the finalizer thread, GC worker threads, ThreadPool worker threads).   This would be very helpful (and also pretty easy).

Vance

**Comment by danmosemsft** [comment]
@jkotas do we have a way to use RS2 specific API in the clr?

**Comment by jkotas** [comment]
Yes: LoadLibrary/GetProcAddress. There is a lot of prior art. One example from many: https://github.com/dotnet/coreclr/blob/13e7c4368da664a8b50228b1a5ef01a660fbb2dd/src/classlibnative/bcltype/system.cpp#L74

**Comment by danmosemsft** [comment]
Doh ! Yes obviously. @Alois-xx any interest in submitting a PR, at least for Thread.Name if not more? Seems reasonably straightforward.

**Comment by Alois-xx** [comment]
Sure. I will get the things and submit a PR. But it can take a while until my limited free has enough gaps.

**Comment by danmosemsft** [comment]
@Alois-xx do you still plan to give this a shot?

/cc @safern

**Comment by Alois-xx** [comment]
Sure there is already a PR pending. 
https://github.com/dotnet/coreclr/pull/12593
So far it should be ok except that I could not add the thread names to GC threads. I think I would need to add some include directory to the GC project but I was not yet able to decipher how the build system works. It is a pretty complex mixture of generated files and text processing as far as I can tell.

**Comment by danmosemsft** [comment]
@Alois-xx if you believe there are more GC threads to name that you did not handle in https://github.com/dotnet/coreclr/pull/12593 then check with @swgillespie
